### PR TITLE
- added flexiblegmres

### PR DESCRIPTION
**Pull request by hnil**
- added flexgmres

### CODE REVIEW

**Review by akva2 — COMMENTED** [review]
The (c&p inherited) lack of std::make_shared usage is a possible improvement.

**Comment by bska** [comment]
jenkins build this please

File: opm/simulators/linalg/FlexibleSolver_impl.hpp [hunk 1]
```diff
@@ -170,6 +170,15 @@ namespace Dune
                                                                         restart, // desired residual reduction factor
                                                                         maxiter, // maximum number of iterations
                                                                         verbosity));
+        } else if (solver_type == "flexgmres") {
+            int restart = prm.get<int>("restart", 15);
+            linsolver_.reset(new Dune::RestartedFlexibleGMResSolver<VectorType>(*linearoperator_for_solver_,
+                                                                                *scalarproduct_,
+                                                                                *preconditioner_,
+                                                                                tol,
+                                                                                restart, // desired residual reduction factor
```
[the remark this answers is not in the document]
Comment (hnil) [reply opm/simulators/linalg/FlexibleSolver_impl.hpp:179]: It is a copy of the above. Should I make a separate pull-request to change all?

[the remark this answers is not in the document]
Comment (hnil) [reply opm/simulators/linalg/FlexibleSolver_impl.hpp]: fixed after rebase.

**Comment by blattms** [comment]
Indeed, there is one closing bracket to many. Seems like I did not check the jenkins build.

**Comment by OPMUSER** [comment]
@hnil I need more information on what this PR does, instead of just "added flexgmres", if you want to include this in the release notes.

If you cannot not provide a decent description of what the PR is doing, then I will not waste my time trying to figure it out, and yes I'm not a happy camper with this PR.

**Comment by atgeirr** [comment]
It is an alternative linear solver, similar to bicgstab [default] and gmres that were already available. I think maybe this paper is the initial reference, https://epubs.siam.org/doi/10.1137/0914028, but @hnil please confirm.

**Comment by OPMUSER** [comment]
> It is an alternative linear solver, similar to bicgstab [default] and gmres that were already available. I think maybe this paper is the initial reference, https://epubs.siam.org/doi/10.1137/0914028, but @hnil please confirm.

@hnil Any body home?

**Pull request merged**